File: Battleship.java
// Battleship.java
// Версия на Java с record, enum, Stream API, Scanner

import java.util.*;
import java.util.regex.Pattern;

public class Battleship {

    // ANSI-цвета (поддерживаются не везде, но для красоты)
    private static final String RESET = "\033[0m";
    private static final String RED = "\033[91m";
    private static final String GREEN = "\033[92m";
    private static final String BLUE = "\033[94m";
    private static final String YELLOW = "\033[93m";
    private static final String CYAN = "\033[96m";

    private static String colorize(String text, String color) {
        return color + text + RESET;
    }

    // Класс для координат
    public record Coord(int row, int col) {}

    // Класс корабля
    public static class Ship {
        private final List<Coord> cells;
        private final boolean[] hits;

        public Ship(List<Coord> cells) {
            this.cells = cells;
            this.hits = new boolean[cells.size()];
        }

        public boolean isSunk() {
            for (boolean h : hits) if (!h) return false;
            return true;
        }

        public boolean hit(Coord coord) {
            for (int i = 0; i < cells.size(); i++) {
                if (cells.get(i).equals(coord)) {
                    hits[i] = true;
                    return true;
                }
            }
            return false;
        }

        public List<Coord> getCells() { return cells; }
    }

    // Игровое поле
    public static class Board {
        public static final int SIZE = 10;
        private static final int[] SHIP_SIZES = {4, 3, 3, 2, 2, 2, 1, 1, 1, 1};

        private final char[][] grid = new char[SIZE][SIZE];
        private final List<Ship> ships = new ArrayList<>();
        private final Set<Coord> shots = new HashSet<>();

        public Board() {
            for (int r = 0; r < SIZE; r++) {
                Arrays.fill(grid[r], '~');
            }
        }

        public void placeShips() {
            Random rand = new Random();
            for (int size : SHIP_SIZES) {
                boolean placed = false;
                int attempts = 0;
                while (!placed && attempts < 1000) {
                    attempts++;
                    int row = rand.nextInt(SIZE);
                    int col = rand.nextInt(SIZE);
                    boolean horizontal = rand.nextBoolean();
                    List<Coord> cells = getCells(row, col, size, horizontal);
                    if (cells != null && canPlace(cells)) {
                        placeShip(cells);
                        placed = true;
                    }
                }
                if (!placed) throw new RuntimeException("Не удалось разместить корабли");
            }
        }

        private List<Coord> getCells(int row, int col, int size, boolean horizontal) {
            List<Coord> cells = new ArrayList<>();
            for (int i = 0; i < size; i++) {
                int r = horizontal ? row : row + i;
                int c = horizontal ? col + i : col;
                if (r >= SIZE || c >= SIZE) return null;
                cells.add(new Coord(r, c));
            }
            return cells;
        }

        private boolean canPlace(List<Coord> cells) {
            for (Coord coord : cells) {
                int r = coord.row(), c = coord.col();
                if (grid[r][c] != '~') return false;
                for (int dr = -1; dr <= 1; dr++) {
                    for (int dc = -1; dc <= 1; dc++) {
                        int nr = r + dr, nc = c + dc;
                        if (nr >= 0 && nr < SIZE && nc >= 0 && nc < SIZE) {
                            if (grid[nr][nc] != '~') return false;
                        }
                    }
                }
            }
            return true;
        }

        private void placeShip(List<Coord> cells) {
            Ship ship = new Ship(cells);
            ships.add(ship);
            for (Coord coord : cells) {
                grid[coord.row()][coord.col()] = '#';
            }
        }

        public String receiveShot(Coord coord) {
            int r = coord.row(), c = coord.col();
            if (r < 0 || r >= SIZE || c < 0 || c >= SIZE) return "invalid";
            if (shots.contains(coord)) return "already_shot";
            shots.add(coord);

            if (grid[r][c] == '#') {
                for (Ship ship : ships) {
                    if (ship.hit(coord)) {
                        grid[r][c] = 'X';
                        if (ship.isSunk()) {
                            for (Coord sc : ship.getCells()) {
                                grid[sc.row()][sc.col()] = 'X';
                            }
                            return "sunk";
                        }
                        return "hit";
                    }
                }
            } else {
                grid[r][c] = 'O';
                return "miss";
            }
            return "miss";
        }

        public boolean allShipsSunk() {
            return ships.stream().allMatch(Ship::isSunk);
        }

        public void display(boolean hideShips) {
            System.out.print("  ");
            for (int i = 1; i <= SIZE; i++) System.out.print(i + " ");
            System.out.println();
            for (int r = 0; r < SIZE; r++) {
                char rowLabel = (char) ('A' + r);
                System.out.print(rowLabel + " ");
                for (int c = 0; c < SIZE; c++) {
                    char cell = grid[r][c];
                    if (hideShips && cell == '#') cell = '~';
                    String colored;
                    switch (cell) {
                        case 'X': colored = colorize("X", RED); break;
                        case 'O': colored = colorize("O", BLUE); break;
                        case '#': colored = colorize("#", GREEN); break;
                        default: colored = String.valueOf(cell);
                    }
                    System.out.print(colored + " ");
                }
                System.out.println();
            }
        }

        public Set<Coord> getShots() { return shots; }
    }

    // Игра
    public static class Game {
        private final Board playerBoard = new Board();
        private final Board computerBoard = new Board();
        private boolean playerTurn = true;
        private boolean gameOver = false;
        private final Scanner scanner = new Scanner(System.in);

        public Game() {
            playerBoard.placeShips();
            computerBoard.placeShips();
        }

        private Coord parseCoordinate(String input) {
            Pattern pattern = Pattern.compile("^([A-Ja-j])([1-9]|10)$");
            var matcher = pattern.matcher(input);
            if (!matcher.matches()) return null;
            int row = matcher.group(1).toUpperCase().charAt(0) - 'A';
            int col = Integer.parseInt(matcher.group(2)) - 1;
            return new Coord(row, col);
        }

        private void playerMove() {
            System.out.println("\nВаше поле:");
            playerBoard.display(false);
            System.out.println("\nПоле противника:");
            computerBoard.display(true);

            while (true) {
                System.out.print("Введите координату (например, A1): ");
                String input = scanner.nextLine().trim();
                Coord coord = parseCoordinate(input);
                if (coord == null) {
                    System.out.println("Неверный формат. Используйте букву A-J и цифру 1-10.");
                    continue;
                }
                String result = computerBoard.receiveShot(coord);
                switch (result) {
                    case "invalid": System.out.println("Координата вне поля."); break;
                    case "already_shot": System.out.println("Сюда уже стреляли."); break;
                    case "hit": System.out.println(colorize("Попадание!", GREEN)); break;
                    case "sunk": System.out.println(colorize("Корабль потоплен!", YELLOW)); break;
                    case "miss": System.out.println(colorize("Промах.", BLUE)); break;
                }
                if (!result.equals("invalid") && !result.equals("already_shot")) break;
            }

            if (computerBoard.allShipsSunk()) {
                System.out.println(colorize("Поздравляем! Вы потопили все корабли противника!", GREEN));
                gameOver = true;
            }
        }

        private void computerMove() {
            System.out.println("\nХод компьютера...");
            Random rand = new Random();
            Coord coord;
            do {
                int row = rand.nextInt(Board.SIZE);
                int col = rand.nextInt(Board.SIZE);
                coord = new Coord(row, col);
            } while (playerBoard.getShots().contains(coord));

            String result = playerBoard.receiveShot(coord);
            char rowChar = (char) ('A' + coord.row());
            String coordStr = rowChar + String.valueOf(coord.col() + 1);
            if (result.equals("hit") || result.equals("sunk")) {
                System.out.println("Компьютер попал в " + coordStr + "!");
            } else {
                System.out.println("Компьютер промахнулся по " + coordStr + ".");
            }

            if (playerBoard.allShipsSunk()) {
                System.out.println(colorize("Компьютер уничтожил все ваши корабли. Вы проиграли.", RED));
                gameOver = true;
            }
        }

        public void run() {
            System.out.println(colorize("Добро пожаловать в Морской бой!", CYAN));
            while (!gameOver) {
                if (playerTurn) playerMove();
                else computerMove();
                playerTurn = !playerTurn;
            }
        }
    }

    public static void main(String[] args) {
        new Game().run();
    }
}

File: battleship.js
// battleship.js
// Версия на JavaScript с использованием readline, классов, async/await

const readline = require('readline');

const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

// Цвета для терминала (ANSI)
const colors = {
    reset: '\x1b[0m',
    red: '\x1b[91m',
    green: '\x1b[92m',
    blue: '\x1b[94m',
    yellow: '\x1b[93m',
    cyan: '\x1b[96m'
};

function colorize(text, color) {
    return `${colors[color] || ''}${text}${colors.reset}`;
}

class Ship {
    constructor(cells) {
        this.cells = cells; // массив [row, col]
        this.hits = new Array(cells.length).fill(false);
    }

    isSunk() {
        return this.hits.every(h => h);
    }

    hit(row, col) {
        for (let i = 0; i < this.cells.length; i++) {
            const [r, c] = this.cells[i];
            if (r === row && c === col) {
                this.hits[i] = true;
                return true;
            }
        }
        return false;
    }
}

class Board {
    static SIZE = 10;
    static SHIP_SIZES = [4, 3, 3, 2, 2, 2, 1, 1, 1, 1];

    constructor() {
        this.grid = Array.from({ length: Board.SIZE }, () => Array(Board.SIZE).fill('~'));
        this.ships = [];
        this.shots = new Set(); // строки "row,col"
    }

    placeShips() {
        for (const size of Board.SHIP_SIZES) {
            let placed = false;
            let attempts = 0;
            while (!placed && attempts < 1000) {
                attempts++;
                const row = Math.floor(Math.random() * Board.SIZE);
                const col = Math.floor(Math.random() * Board.SIZE);
                const horizontal = Math.random() < 0.5;
                const cells = this._getCells(row, col, size, horizontal);
                if (cells && this._canPlace(cells)) {
                    this._placeShip(cells);
                    placed = true;
                }
            }
            if (!placed) throw new Error('Не удалось разместить корабли');
        }
    }

    _getCells(row, col, size, horizontal) {
        const cells = [];
        for (let i = 0; i < size; i++) {
            const r = horizontal ? row : row + i;
            const c = horizontal ? col + i : col;
            if (r >= Board.SIZE || c >= Board.SIZE) return null;
            cells.push([r, c]);
        }
        return cells;
    }

    _canPlace(cells) {
        for (const [r, c] of cells) {
            if (this.grid[r][c] !== '~') return false;
            for (let dr = -1; dr <= 1; dr++) {
                for (let dc = -1; dc <= 1; dc++) {
                    const nr = r + dr, nc = c + dc;
                    if (nr >= 0 && nr < Board.SIZE && nc >= 0 && nc < Board.SIZE) {
                        if (this.grid[nr][nc] !== '~') return false;
                    }
                }
            }
        }
        return true;
    }

    _placeShip(cells) {
        const ship = new Ship(cells);
        this.ships.push(ship);
        for (const [r, c] of cells) {
            this.grid[r][c] = '#';
        }
    }

    receiveShot(row, col) {
        if (row < 0 || row >= Board.SIZE || col < 0 || col >= Board.SIZE) return 'invalid';
        const key = `${row},${col}`;
        if (this.shots.has(key)) return 'already_shot';
        this.shots.add(key);

        if (this.grid[row][col] === '#') {
            for (const ship of this.ships) {
                if (ship.hit(row, col)) {
                    this.grid[row][col] = 'X';
                    if (ship.isSunk()) {
                        for (const [r, c] of ship.cells) {
                            this.grid[r][c] = 'X';
                        }
                        return 'sunk';
                    }
                    return 'hit';
                }
            }
        } else {
            this.grid[row][col] = 'O';
            return 'miss';
        }
        return 'miss';
    }

    allShipsSunk() {
        return this.ships.every(ship => ship.isSunk());
    }

    display(hideShips = false) {
        console.log('  ' + Array.from({ length: Board.SIZE }, (_, i) => i + 1).join(' '));
        for (let r = 0; r < Board.SIZE; r++) {
            const rowLabel = String.fromCharCode(65 + r);
            let rowCells = [];
            for (let c = 0; c < Board.SIZE; c++) {
                let cell = this.grid[r][c];
                if (hideShips && cell === '#') cell = '~';
                let colored = cell;
                if (cell === 'X') colored = colorize('X', 'red');
                else if (cell === 'O') colored = colorize('O', 'blue');
                else if (cell === '#') colored = colorize('#', 'green');
                rowCells.push(colored);
            }
            console.log(rowLabel + ' ' + rowCells.join(' '));
        }
    }
}

class BattleshipGame {
    constructor() {
        this.playerBoard = new Board();
        this.computerBoard = new Board();
        this.playerBoard.placeShips();
        this.computerBoard.placeShips();
        this.playerTurn = true;
        this.gameOver = false;
    }

    parseCoordinate(coord) {
        const match = coord.match(/^([A-Ja-j])([1-9]|10)$/);
        if (!match) return null;
        const row = match[1].toUpperCase().charCodeAt(0) - 65;
        const col = parseInt(match[2], 10) - 1;
        return [row, col];
    }

    async playerMove() {
        console.log('\nВаше поле:');
        this.playerBoard.display(false);
        console.log('\nПоле противника:');
        this.computerBoard.display(true);

        while (true) {
            const coord = await this._question('Введите координату (например, A1): ');
            const parsed = this.parseCoordinate(coord.trim());
            if (!parsed) {
                console.log('Неверный формат. Используйте букву A-J и цифру 1-10.');
                continue;
            }
            const [row, col] = parsed;
            const result = this.computerBoard.receiveShot(row, col);
            if (result === 'invalid') {
                console.log('Координата вне поля.');
            } else if (result === 'already_shot') {
                console.log('Сюда уже стреляли.');
            } else {
                if (result === 'hit') console.log(colorize('Попадание!', 'green'));
                else if (result === 'sunk') console.log(colorize('Корабль потоплен!', 'yellow'));
                else if (result === 'miss') console.log(colorize('Промах.', 'blue'));
                break;
            }
        }

        if (this.computerBoard.allShipsSunk()) {
            console.log(colorize('Поздравляем! Вы потопили все корабли противника!', 'green'));
            this.gameOver = true;
        }
    }

    computerMove() {
        console.log('\nХод компьютера...');
        let row, col;
        do {
            row = Math.floor(Math.random() * Board.SIZE);
            col = Math.floor(Math.random() * Board.SIZE);
        } while (this.playerBoard.shots.has(`${row},${col}`));
        const result = this.playerBoard.receiveShot(row, col);
        const coordStr = String.fromCharCode(65 + row) + (col + 1);
        if (result === 'hit' || result === 'sunk') {
            console.log(`Компьютер попал в ${coordStr}!`);
        } else {
            console.log(`Компьютер промахнулся по ${coordStr}.`);
        }

        if (this.playerBoard.allShipsSunk()) {
            console.log(colorize('Компьютер уничтожил все ваши корабли. Вы проиграли.', 'red'));
            this.gameOver = true;
        }
    }

    _question(prompt) {
        return new Promise(resolve => rl.question(prompt, resolve));
    }

    async run() {
        console.log(colorize('Добро пожаловать в Морской бой!', 'cyan'));
        while (!this.gameOver) {
            if (this.playerTurn) {
                await this.playerMove();
            } else {
                this.computerMove();
            }
            this.playerTurn = !this.playerTurn;
        }
        rl.close();
    }
}

const game = new BattleshipGame();
game.run().catch(console.error);
